pkg/scene/widget_align_system.go

```diff
@@ -0,0 +1,250 @@
+package scene
+
+import (
+	"github.com/reearth/reearth-backend/pkg/id"
+)
+
+// WidgetAlignSystem is the layout structure of any enabled widgets that will be displayed over the scene.
+type WidgetAlignSystem struct {
+	inner WidgetZone
+	outer WidgetZone
+}
+
+// WidgetZone is the structure of each layer (inner and outer) of the align system.
+type WidgetZone struct {
+	left   WidgetSection
+	center WidgetSection
+	right  WidgetSection
+}
+
+// WidgetSection is the structure of each section of the align system.
+type WidgetSection struct {
+	top    WidgetArea
+	middle WidgetArea
+	bottom WidgetArea
+}
+
+// WidgetArea has the widgets and alignment information found in each part area of a section.
+type WidgetArea struct {
+	widgetIds []*id.WidgetID
+	align     string
+}
+
+type WidgetLocation struct {
+	Zone    string
+	Section string
+	Area    string
+}
+
+const (
+	AlignStart  = "start"
+	AlignCenter = "center"
+	AlignEnd    = "end"
+)
+
+var Areas = []string{
+	"top",
+	"middle",
+	"bottom",
+}
+
+var Sections = map[string][]string{
+	"left":   Areas,
+	"center": Areas,
+	"right":  Areas,
+}
+
+var Zones = map[string]map[string][]string{
+	"inner": Sections,
+	"outer": Sections,
+}
+
+func (was *WidgetAlignSystem) FindWidgetIDLocation(wid id.WidgetID) (*int, *WidgetLocation) {
```

Comment:
There is room for reconsideration here.

First, the name could be a bit smarter (FindWidgetIDLocation -> FindWidgetLocation). Second, int doesn't have to be a reference. A zero value will suffice.

And the method itself is not very good. Do we really need to return an index? Why do we need this method?

It is because the logic has been concentrated in the WidgetAlignSystem. If each Area's method is called internally by each Section, then each Zone calls that method internally, and finally the WidgetAlignSystem calls the Zone's method, each would only have to worry about the one below it. However, currently, the WidgetAlignSystem takes care of all Areas, and the contents of the Areas are leaking out.

For example, if we had such a method, each would not have to go to the private fields of other structs.

Message passing, as it is often called in object-oriented programming, is exactly what it is. You interact with information via methods.

```go
func (*WidgetAlignSystem) Zone(string) *WidgetZone
func (*WidgetAlignSystem) Find(...) int
func (*WidgetAlignSystem) Widtets() []id.Widget

func (*WidgetZone) Section(string) *WidgetSection
func (*WidgetZone) Find(...) int
func (*WidgetZone) Widtets() []id.Widget

func (*WidgetSection) Area(string) *WidgetArea
func (*WidgetSection) Find(...) int
func (*WidgetSection) Widtets() []id.Widget

func (*WidgetArea) Add(...)
func (*WidgetArea) Update(...)
func (*WidgetArea) Remove(...)
func (*WidgetArea) Find(...) int
func (*WidgetArea) Widtets() []id.Widget
```

Reply:
So, this method is necessary after your comment about the Remove and Update methods not needing a Location param, just the WidgetID. Without the Location, we need to search the whole align system for the widgetID. This was the best method I could think of to search the whole tree. 
Also, the index of the found widget is necessary for both the Remove and Update methods(Remove needs to cut out the single widgetID, and Update possibly will be moving the WidgetID's index. I know that could be handled separately in each method, but it would be duplicated code).

I am a little confused with your suggestion though. I understand the message passing, but for each Find method they return an int. What would the Find method be finding? Ultimately, how does that help with searching the whole align system for a widgetID?

Reply:
First of all, the process of finding a widget in a WidgetAlignSystem is very fast: the CPU is extremely fast, all the data is in memory, and the amount of data is only in the order of 100 at most. You don't need to worry about it.

Since we are assuming that the same widget ID does not exist in more than one place, we don't even need an index number to delete it; we can just give it an ID, search all the areas, and delete it when we find it. The logic can be hidden in the respective Remove methods.

- Remove method of WidgetAlignSystem calls the Remove methods of the two Zones.
- Remove method of Zone calls the Remove methods of the three Sections.
- Remove method of Section calls the three Area Remove methods.
- Remove method of Area loops through the slice with the IDs, and removes the ID when the IDs match.

The key point of this implementation is that WidgetAlignSystem does not manipulate or reference the fields of the Area directly; WidgetAlignSystem has no knowledge of the internals of Area and manipulates them via methods.

Since the number of data is small, this implementation is perfectly fine, and more importantly, it dramatically improves the readability of the code. If you are really concerned about the extra loop, you can make each Remove method return true when the ID is found, and break all loops when it is found.